packages/flame/lib/src/components/fps_component.dart

```diff
@@ -0,0 +1,31 @@
+import 'dart:collection';
+
+import '../../components.dart';
+
+/// The [FPSComponent] is a non-visual component which you can get the current
+/// fps of the game with by calling [fps], once the component has been added to
+/// the component tree.
+class FPSComponent extends Component {
+  FPSComponent({
+    this.windowSize = 60,
+  });
+
+  final int windowSize;
+  final Queue<double> _window = Queue();
+  double _sum = 0;
+
+  @override
+  void update(double dt) {
+    final frameFps = 1 / dt;
+    _window.add(frameFps);
+    _sum += frameFps;
+    if (_window.length > windowSize) {
+      _sum -= _window.removeFirst();
+    }
+  }
+
+  /// Get the current average FPS over the last [windowSize] frames.
+  double get fps {
+    return _window.isEmpty ? 0 : _sum / _window.length;
+  }
```

Comment:
I recommend averaging times `dt`, not the `fps` values. Consider, for example, what would happen if there is a single `dt` that is very high, or very low.

```suggestion
  void update(double dt) {
    _window.addLast(dt);
    _sum += dt;
    if (_window.length > windowSize) {
      _sum -= _window.removeFirst();
    }
  }

  /// Get the current average FPS over the last [windowSize] frames.
  double get fps {
    return _window.isEmpty ? 0 : _window.length / _sum;
  }
```

Reply:
I did that from the beginning, but the rounding errors become quite big.

```
So say that we do 4 frames, 2 of them has the dt for 40fps and 2 for 60fps:
40fps = 1/40 = 0.025dt
60fps = 1/60 = 0.016...dt
(0.025+0.025+0.016+0.016)/4=0.0208... (average dt)
1/0.0208=48fps (dt to fps)
(Should be 50)
```

Reply:
The average of 40fps and 60fps is `2/(1/40 + 1/60) = 48 fps`.
"FPS" literally means "frames per second", i.e. how many frames there were (on average) in one second.

Imagine you have one frame rendering in 0.999s, and another frame rendering in 0.001s. Would you say it's 2 frames per second; or is it (1.001 + 1000)/2 ≈ 500 fps?

Reply:
That's true, late Sunday evenings clearly doesn't produce the best code. 😄